src/Confirmation.vala

```diff
@@ -19,28 +19,57 @@
 */
 
 public class Notifications.Confirmation : AbstractBubble {
+    public string confirmation_type { get; construct set; }
     public new string icon_name { get; construct set; }
     public double progress { get; construct set; }
 
-    public Confirmation (string icon_name, double progress) {
+    private Gtk.Image image;
+    private Gtk.ProgressBar progressbar;
+
+    public Confirmation (string confirmation_type, string icon_name, double progress) {
         Object (
+            confirmation_type: confirmation_type,
             icon_name: icon_name,
             progress: progress,
             timeout: 2000
         );
     }
 
     construct {
-        var image = new Gtk.Image.from_icon_name (icon_name, Gtk.IconSize.DIALOG) {
+        var contents = create_contents ();
+        content_area.add (contents);
+
+        get_style_context ().add_class ("confirmation");
+    }
+
+    public void replace (string confirmation_type, string icon_name, double progress) {
+        if (this.confirmation_type == confirmation_type) {
+            image.icon_name = icon_name;
+            progressbar.fraction = progress;
+            return;
+        }
+
+        this.confirmation_type = confirmation_type;
+        this.icon_name = icon_name;
+        this.progress = progress;
+
+        var contents = create_contents ();
+        content_area.add (contents);
+        content_area.visible_child = contents;
+    }
```

Comment:
an alternative way do to this, is, connecting to `notify::confirmation-type` (or even using the set method, like is done in the normal bubble), and binding the `icon-name` and `progress` property to the contents `Gtk.Image:icon-name` and `Gtk.ProgressBar:fraction`.

this way you don't need to store the progress bar and image and only need to deal with the case that confirmation-type is changed.

Reply:
But we need to change all three parameters all at once, so that the old confirmation smoothly transitions into the new one. Or am I missing something here?

Reply:
visual changes will only have effect in the next iteration of the global context, since signals are synchronous, and bindings use signals under the hood. everything will be updated before Gtk start to show the new contents.